---
title: "Decontamination of BT 2023 Rockfish Amplicons - after assigning ASVs a taxa"
author: "Kimberly Ledger"
output: html_document
date: "2025-03-11"
---

libraries
```{r}
library(tidyverse)
library(stringr)
library(reshape2)
rename <- dplyr::rename
```

load sample metadata
```{r}
metadata <- read.csv("/home/kimberly.ledger/goa-bt23-mb/data/GOA_BT2023_metadata.csv") %>%
  filter(amplicon == "rkfish")
```

check sequence table outputs
```{r}
asv_table <- readRDS("/home/kimberly.ledger/goa-bt23-mb/data/dadasnake_rkfish/filtered.seqTab.RDS") %>%
  select(!Row.names)

#transpose 
asv_table <- data.frame(t(asv_table))

#set column names to be ASV# 
colnames(asv_table) <- asv_table["ASV",]

#remove row that has ASV#
asv_table <- asv_table[!rownames(asv_table) %in% c('ASV'), ]

#make sure reads are numbers
# Convert all character columns to numeric
for (col in names(asv_table)) {
  asv_table[[col]] <- as.numeric(asv_table[[col]])
}

asv_table$sample_ID <- rownames(asv_table)  #make make sample ID a column 
asv_table$sample_ID <- gsub("-", "_", asv_table$sample_ID) #illumina output changed "_" to "-"

#reorder columns and change NTC names to remove extra "_"
asv_table <- asv_table %>%
  select(sample_ID, everything()) %>%
  mutate(sample_ID = str_replace(sample_ID, "NTC_", "NTC"))
```

add column to the ASV table that labels the sample type
```{r}
asv_table_with_sample_type <- metadata %>%
  dplyr::select(sample_ID, sample_type) %>%
  left_join(asv_table, by = "sample_ID") %>%
  mutate(across(everything(), ~ replace_na(.x, 0)))

# make a variable for the first and last ASV column in the table
asv_first <- which(colnames(asv_table_with_sample_type) == "ASV_0923")  #this is the first asv column in this dataframe
asv_last <- ncol(asv_table_with_sample_type)
```

pivot table longer 
```{r}
asv_table_long <- asv_table_with_sample_type %>%
  pivot_longer(cols = c(asv_first:asv_last), names_to = "ASV", values_to = "reads") %>%
  mutate(reads = as.numeric(reads)) %>%
  mutate(reads = ifelse(is.na(reads), 0, reads))

length(unique(asv_table_long$ASV))
```

## Step 1. Remove ASVs that don't get a taxonomic assignment 
```{r}
taxonomy <- read.csv("/home/kimberly.ledger/goa-bt23-mb/outputs/rkfish_taxonomy_20250130_collapsed.csv") %>% 
  select(!X) %>%
  rename(ASV = qseqid)

asv_table_filter1 <- asv_table_long %>%
  filter(ASV %in% taxonomy$ASV)

length(unique(taxonomy$ASV))
```

## convert ASV table to taxon table 
```{r}
taxon_table <- asv_table_filter1 %>%
  left_join(taxonomy) %>%
  group_by(across(-c(ASV, reads))) %>%
  summarize(tot_reads = sum(reads)) %>%
  rename(reads = tot_reads)
```

### next... make some plots

pcr_blanks 
```{r}
taxon_table %>%
  #filter(reads > 0) %>%
  filter(sample_type == "pcr_blank") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
    theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in PCR blanks") + 
  theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

```{r}
taxon_pcrblank <- taxon_table %>%
  filter(reads > 0) %>%
  filter(sample_type == "pcr_blank") %>%
  group_by(taxon) %>%
  summarise(total = sum(reads)) %>%
  arrange(desc(total))
```

some reads in pcr blanks 

extraction_blanks 
```{r}
taxon_table %>%
  #filter(reads > 0) %>%
  filter(sample_type == "extraction_blank") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
    theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in extraction blanks") + 
 theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

```{r}
taxon_extractionblank <- taxon_table %>%
  filter(reads > 0) %>%
  filter(sample_type == "extraction_blank") %>%
  group_by(taxon) %>%
  summarise(total = sum(reads)) %>%
  arrange(desc(total))
```

quite a few reads in some of the extraction blank replicates

field_blanks 
```{r}
taxon_table %>%
  #filter(reads > 0) %>%
  filter(sample_type == "field_blank") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
    theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field blanks") + 
 theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

it'd be interesting to compare this to the mifish amp of the field blanks - there didn't seem to be too much Sebastes there but i could be remembering wrong 

```{r}
taxon_fieldblank <- taxon_table %>%
  filter(reads > 0) %>%
  filter(sample_type == "field_blank") %>%
  group_by(taxon) %>%
  summarise(total = sum(reads)) %>%
  arrange(desc(total))
```

positive 
```{r}
taxon_table %>%
  #filter(reads > 0) %>%
  filter(sample_type == "positive") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in positive") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

as expected... SOO many reads for these... 

any tag-jumping?
```{r}
prop_asvs_in_positives <- taxon_table %>%
  filter(sample_type == "positive") %>%
  group_by(sample_ID) %>%
  mutate(TotalReadsPerSample = sum(reads)) %>%
  mutate(Prop = reads/TotalReadsPerSample) %>%
  group_by(taxon) %>%
  summarise(max_prop = max(Prop)) %>%
  arrange(desc(max_prop))
```

very very little % of other spp in positive 

field samples 
```{r}
taxon_table %>%
  #filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "rockfish reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

```{r}
taxon_table %>%
  filter(sample_type == "sample") %>%
  group_by(sample_ID) %>%
  summarize(total_reads = sum(reads)) %>%
  mutate(reads_bin = case_when(
    total_reads == 0 ~ "0",
    total_reads > 0 & total_reads < 100 ~ "1-99",
    total_reads >= 100 & total_reads < 500 ~ "100-499",
    total_reads >= 500 & total_reads < 1000 ~ "500-999",
    total_reads >= 1000 & total_reads < 2000 ~ "1000-1999",
    total_reads >= 2000 ~ ">2000"
  )) %>%
  count(reads_bin)

sample_read_count <- taxon_table %>%
  group_by(sample_ID) %>%
  summarize(total_reads = sum(reads)) %>%
  separate(sample_ID, into = c("id", "replicate", "assay"), sep = "_") %>%
  arrange(replicate)

#write.csv(sample_read_count, "~/goa-bt23-mb/outputs/rkfish_read_count.csv")
```



```{r}
taxon_fieldsamples <- taxon_table %>%
  filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  group_by(taxon) %>%
  summarise(total = sum(reads)) %>%
  arrange(desc(total))
```

VERY spotty amplification... but when there is amplification, it generally looks like it's all from one taxa (which is probably a good thing.)
- pop as most reads makes sense
- odd that the Sebastes emphaeus_variegatus_wilsoni group is second?? only S. variegatus is in trawl catch 

zoom in on a few drops 
```{r}
metadata_haul <- metadata %>%
  select(sample_ID, extraction_ID, haul)

taxon_table %>%
  left_join(metadata_haul) %>%
  filter(haul == 1) %>%
  #filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  facet_wrap(sample_type~extraction_ID, scales = 'free', ncol = 4) +
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

S. alutus in one or two technical reps per biological rep. 


```{r}
taxon_table %>%
  left_join(metadata_haul) %>%
  filter(haul == 2) %>%
  #filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  facet_wrap(sample_type~extraction_ID, scales = 'free', ncol = 3) +
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

two bio reps with very low read counts and various species, two bioreps with high read count and pop and Sebastes melanops


compare Sebastes d-loop reads to 
```{r}
seb_mifish <- read.csv("/home/kimberly.ledger/goa-bt23-mb/outputs/sebastes_reads_in_mifish.csv") %>%
  select(!X) %>%
  separate(sample_ID, into = c("extraction_ID", "replicate", "amplicon"), sep = "_", remove = F) %>%
  select(!amplicon) %>%
  filter(total_seb > 100) %>%   ## pcr rep need at least 100 rockfish reads
  group_by(extraction_ID) %>%
  summarize(seb_reads = sum(total_seb),
            seb_reps = n_distinct(sample_ID))
```

join to taxon table
```{r}
taxon_table <- taxon_table %>%
  #filter(sample_type != "positive") %>%
  #filter(sample_type != "pcr_blank") %>%
  separate(sample_ID, into = c("extraction_ID", "replicate", "amplicon"), sep = "_", remove = F) %>%
  select(!amplicon) %>%
  mutate(rkfish_in_mifish = ifelse(extraction_ID %in% seb_mifish$extraction_ID, "yes", "no"))
```


field samples that mifish mb detected rockfish
```{r}
taxon_table %>%
  filter(rkfish_in_mifish == "yes") %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

```{r}
taxon_table %>%
  filter(rkfish_in_mifish == "no") %>%
  filter(sample_type == "sample") %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```


zoom in on samples w/ rockfish reads but few 
```{r}
taxon_table %>%
  filter(sample_type == "sample") %>%
  group_by(sample_ID) %>%
  mutate(Tot = sum(reads)) %>%
  filter(Tot > 100 & Tot < 1000) %>%
  ggplot(aes(x=sample_ID, y=reads, fill=taxon)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  labs(
    y = "number of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 6),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "bottom",
    #legend.position = "none",
    legend.title = element_blank()
  )
```

```{r}
taxon_table %>%
  filter(sample_type == "sample") %>%
  group_by(sample_ID) %>%
  mutate(Tot = sum(reads)) %>%
  filter(Tot > 100 & Tot < 1000) %>%
  group_by(taxon) %>%
  summarize(tot_reads = sum(reads)) %>%
  arrange(desc(tot_reads))
```



let me see what rockfish species show up in the bottom trawls 

```{r}
sebastes <- read.csv("/home/kimberly.ledger/goa-bt23-mb/data/CATCH_AND_HAUL_DATA_GOA_BT_2023_Chordata.csv") %>%
  filter(str_detect(Taxon.scientific.name, "^Sebast"))
```

```{r}
sebastes %>%
  group_by(Taxon.common.name, Taxon.scientific.name) %>%
  summarize(CPUE_kg = sum(Weight.CPUE..kg.km2.),
            CPUE_n = sum(Number.CPUE..no.km2.)) %>%
  arrange(desc(CPUE_kg))
```


identify species in both eDNA and trawl
```{r}
both_lists <- intersect(taxon_table$taxon, sebastes$Taxon.scientific.name)
both_lists
```

identify species in eDNA but not trawl
```{r}
eDNA_only <- setdiff(taxon_table$taxon, sebastes$Taxon.scientific.name)
eDNA_only
```

S. atrovirens not in range
S. carnatus not in range - used as positive control so may need to account for tag jumping??? 
S. diaconus not usually in GOA 
some others not really in the range... check read numbers 


identify species in trawl but not eDNA
```{r}
trawl_only <- setdiff(sebastes$Taxon.scientific.name, taxon_table$taxon)
trawl_only
```

*check S. altivelis and macrochir in reference db, those and S. paucispinis and S. diploproa not detected in eDNA 


## Step 2. check out taxa that do not occur in field samples
```{r}
reads_per_type_taxa <- taxon_table %>%
  group_by(taxon, sample_type) %>%
  summarize(TotalReadsPerTaxa = sum(reads, na.rm = TRUE)) %>%
  arrange(taxon)
```

what ASVs have no reads in samples, but reads in the controls? 
```{r}
not_in_samples <- reads_per_type_taxa %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerTaxa")) %>%
    filter(sample < 1)
not_in_samples
```

these all make sense to remove

what ASVs do have reads in samples, but more reads in the controls? 
```{r}
more_in_pcr_blanks <- reads_per_type_taxa %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerTaxa")) %>%
  #filter(sample > 1) %>%
  filter(pcr_blank > sample)
more_in_pcr_blanks

more_in_extraction_blanks <- reads_per_type_taxa %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerTaxa")) %>%
  #filter(sample > 1) %>%
  filter(extraction_blank > sample)
more_in_extraction_blanks

more_in_fb_blanks <- reads_per_type_taxa %>%
  pivot_wider(names_from = "sample_type", values_from = c("TotalReadsPerTaxa")) %>%
  #filter(sample > 1) %>%
  filter(field_blank > sample)
more_in_fb_blanks
```




```{r}
metadata_haul <- metadata %>%
  select(sample_ID, extraction_ID, haul, Bottom_type, Average_depth) %>%
     mutate(Bottom_type = ifelse(Bottom_type == "T", "trawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UT", "untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNK", "trawlable and untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKT", "trawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKUT", "untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKTUT", "trawlable and untrawlable", Bottom_type))

# Create a 21-color palette (modify as needed)
palette_21 <- c(
  "#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", 
  "#8c564b", "#e377c2", "#7f7f7f", "#bcbd22", "#17becf",
  "#aec7e8", "#ffbb78", "#98df8a", "#ff9896", "#c5b0d5", 
  "#c49c94", "#f7b6d2",  "#c7c7c7", "#dbdb8d", "#9edae5", "#00736D")

rkfish_fig <- taxon_table %>%
  left_join(metadata_haul) %>%
  filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  group_by(sample_ID) %>%
  mutate(tot = sum(reads),
         prop = reads/tot) %>%
  filter(tot > 2000)  %>%
  filter(sample_ID != "e03802_B_rkfish") %>% 
  ggplot(aes(x=sample_ID, y=prop, fill=taxon)) +
  geom_bar(stat = "identity") + 
  scale_fill_manual(values = palette_21) +
  facet_wrap(haul~Bottom_type~Average_depth, scales = 'free', ncol = 5) +
  theme_bw() +
  labs(
    y = "proportion of sequencing reads",
    x = "sample ID",
    title = "taxon reads in field samples") + 
theme(
    #axis.text.x = element_text(angle = 90, hjust = 0.95),
    axis.text.x = element_blank(),
    legend.text = element_text(size = 9),
    legend.key.size = unit(0.3, "cm"),
    legend.position = "right",
    #legend.position = "none",
    legend.title = element_blank()
  ) + 
  guides(fill = guide_legend(ncol = 1))
rkfish_fig 

rkfish_fig %>% select(sample_ID, tot) %>% unique() %>% arrange(desc(tot))
```
```{r}
#ggsave("/home/kimberly.ledger/goa-bt23-mb/figures/rkfish.png", plot = rkfish_fig, width = 12, height = 6, dpi = 300)
```


```{r}
rk_heatmap <- taxon_table %>%
  left_join(metadata_haul) %>%
  filter(Bottom_type != "trawlable and untrawlable") %>%
  filter(reads > 0) %>%
  filter(sample_type == "sample") %>%
  group_by(sample_ID) %>%
  mutate(tot = sum(reads),
         prop = reads/tot) %>%
  filter(tot > 2000)  %>%
  filter(sample_ID != "e03802_B_rkfish") %>%
  filter(prop > 0.1) %>%
  select(taxon, haul, Bottom_type) %>%
  ungroup() %>%
  unique() %>%
  count(taxon, Bottom_type) %>%
  mutate(n_bin = ifelse(n > 0, 1, 0)) %>%
  ggplot(aes(x = Bottom_type, y = taxon, fill = Bottom_type)) +
  geom_tile(color = "white") +
  scale_fill_manual(values = c("trawlable" = "#F8766D", "untrawlable" = "#00BFC4")) +
  theme_classic() +
  labs(x = "bottom type", 
       y = "taxon") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none")

rk_heatmap 
```

```{r}
ggsave("/home/kimberly.ledger/goa-bt23-mb/figures/rkfish_heatmap.png", plot = rk_heatmap, width = 6, height = 6, dpi = 300)
```


## Step 3. Explore low read depth samples based on taxa accumulation curve

```{r}
library(vegan)

taxon_table_wide <- taxon_table %>%
  ungroup() %>%
  select(sample_ID, taxon, reads) %>%
  mutate(reads = as.integer(reads)) %>%
  pivot_wider(names_from = taxon, values_from = reads)

sample_IDs <- taxon_table_wide$sample_ID

taxon_table_wide <- taxon_table_wide %>%
  ungroup() %>%
  select(-sample_ID)

## plots the figure
rarecurve(taxon_table_wide, step = 20, col = "blue", label = FALSE, 
          main = "Sequencing Effort Curves",
          xlab = "Sequencing Depth", ylab = "Number of Taxa Identified",
          xlim = c(0,1000))
```

the taxa vs seq depth curves plateau at a pretty low read depth (which is good)   

summarize in a table how many pcr replicates meet certain read count thresholds 
```{r}
read_summary <- taxon_table %>%
  group_by(sample_ID, sample_type) %>%
  summarize(tot_reads = sum(reads)) %>%
  arrange(desc(tot_reads)) %>%
  group_by(sample_type) %>%
  summarize(atleast0 = sum(tot_reads >= 0),
            atleast100 = sum(tot_reads >= 100),
            atleast200 = sum(tot_reads >= 200),
            atleast500 = sum(tot_reads >= 500),
            atleast1k = sum(tot_reads >= 1000))
read_summary
```



```{r}
taxon_table_filter <- taxon_table %>%
  filter(!taxon %in% not_in_samples$taxon) %>%
  group_by(sample_ID) %>%
  mutate(Tot = sum(reads)) %>%
  filter(Tot > 2000)
```



summarize read counts by taxa - trawlable vs untrawlable
```{r}
taxon_table_filter %>%
  left_join(metadata_hab, by = "sample_ID") %>%
  filter(sample_type == "sample")  %>%
  filter(Bottom_type == "trawlable") %>%
   separate(sample_ID, into = c("extraction_ID", "replicate", "amplicon"), sep = "_", remove = F) %>%
  filter(reads > 0) %>%
  group_by(taxon) %>%
  summarise(total_reads = sum(reads),
            N_bio_reps = n_distinct(extraction_ID),
            N_pcr_reps = n_distinct(sample_ID)) %>%
  arrange(desc(N_bio_reps))

taxon_table_filter %>%
  left_join(metadata_hab, by = "sample_ID") %>%
  filter(sample_type == "sample")  %>%
  filter(Bottom_type == "untrawlable") %>%
   separate(sample_ID, into = c("extraction_ID", "replicate", "amplicon"), sep = "_", remove = F) %>%
  filter(reads > 0) %>%
  group_by(taxon) %>%
   summarise(total_reads = sum(reads),
            N_bio_reps = n_distinct(extraction_ID),
            N_pcr_reps = n_distinct(sample_ID)) %>%
  arrange(desc(N_bio_reps))
```


#### multivariate plots ##### 

using PCR replicates as individual samples - calculate eDNA index according to Kelly et al. 2019

```{r}
library(vegan)

wide_df <- taxon_table_filter %>%
  mutate(prop = reads/Tot) %>%
  filter(prop > 0.1) %>%
  #filter(reads > 0) %>%
  select(sample_ID, reads, taxon) %>%
  pivot_wider(names_from = taxon, values_from = reads, values_fill = 0)
  
ids <- wide_df$sample_ID
wide_df <- wide_df[,-c(1)]

wis_index <- wisconsin(wide_df)
rowSums(wis_index)

wis_index$sample_ID <- ids

wis_index <- wis_index %>%
  select(sample_ID, everything())

## join the metadata 
metadata_mini <- metadata %>%
  select(sample_ID, haul, start_dd_lat, start_dd_long, depth, Bottom_type) %>%
  mutate(Bottom_type = ifelse(Bottom_type == "T", "trawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UT", "untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNK", "trawlable and untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKT", "trawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKUT", "untrawlable", Bottom_type),
         Bottom_type = ifelse(Bottom_type == "UNKTUT", "trawlable and untrawlable", Bottom_type)) %>%
  mutate(Bottom_type = as.factor(Bottom_type))


wis_index_w_meta <- wis_index %>%
  left_join(metadata_mini) %>%
  filter(Bottom_type != "trawlable and untrawlable") %>%
  select(!`Sebastes flavidus`)  ## only in trawable and untrawable
```



let me try a distance-based redundancy analysis (dbRDA) for continuous variables - latitude and depth
```{r}
wis_df <- wis_index_w_meta[,-c(1,18:22)]

meta_df <- wis_index_w_meta[,c(1,18:22)]


# Fit the dbRDA model - using capscale() - constrained analysis of principal coordinates

#rda_index <- capscale(wis_df ~ + meta_df$start_dd_long + meta_df$depth + as.factor(meta_df$Bottom_type), distance = "jaccard")
rda_index <- capscale(wis_df ~ meta_df$start_dd_long + as.factor(meta_df$Bottom_type), distance = "jaccard")
#rda_index <- capscale(wis_df ~ meta_df$start_dd_long + Condition(as.factor(meta_df$Bottom_type)), distance = "jaccard")
#rda_index <- capscale(wis_df ~ meta_df$start_dd_long, distance = "jaccard")

anova_terms <- anova(rda_index, by = "terms", permutations = 999)
anova_terms 

summary(rda_index)

# Assuming capscale_model is already defined
eigenvalues <- rda_index$CA$eig
total_inertia <- sum(eigenvalues)

# Calculate proportion explained for CAP1 and CAP2
proportion_variance <- eigenvalues / total_inertia * 100

# First use the regular plot option to get a feel for the data
plot(rda_index, scaling = 2, choices = c(1,2))
plot(rda_index, scaling = 2, choices = c(1,3))

# Extract site and species scores
site_scores <- as.data.frame(vegan::scores(rda_index, display = "sites", choices = 1:3))
#site_scores <- as.data.frame(rda_index$CCA$u[, 1:3])
site_scores$Sample <- rownames(site_scores)  # Add sample names

species_scores <- as.data.frame(vegan::scores(rda_index, display = "species", choices = 1:3))
#species_scores <- as.data.frame(rda_index$CCA$v[, 1:3])
species_scores$Species <- rownames(species_scores)  # Add species names

# Extract biplot arrows (for explanatory variables)
biplot_arrows <- as.data.frame(vegan::scores(rda_index, display = "bp", choices = 1:3))
#biplot_arrows <- as.data.frame(rda_index$CCA$biplot[, 1:3])
biplot_arrows$variable <- rownames(biplot_arrows)

biplot_arrows <- biplot_arrows %>%
  mutate(variable = ifelse(variable == "meta_df$start_dd_lat", "latitude", variable),
         variable = ifelse(variable == "meta_df$start_dd_long", "longitude", variable),
         variable = ifelse(variable == "meta_df$depth", "depth", variable),
         variable = ifelse(variable == "as.factor(meta_df$Bottom_type)trawlable and untrawlable", "both", variable),
         variable = ifelse(variable == "as.factor(meta_df$Bottom_type)untrawlable", "untrawlable", variable)) %>%
  filter(variable != "untrawlable")

# use envfit to run signficance test for species 
species_fit <- envfit(rda_index, wis_df, permutations = 999)

# extract significant species
# Get scores and p-values
species_scores_all <- scores(species_fit, display = "vectors") %>% as.data.frame()
species_scores_all$Species <- rownames(species_scores_all)
species_scores_all$pval <- species_fit$vectors$pvals

# Filter to significant species only (e.g., p < 0.05)
species_scores_sig <- species_scores_all %>% 
  filter(pval < 0.2)

library(ggrepel)
## Plot CAP1 vs CAP2 - color sites by bottom type
plot_bottom <- ggplot() +
  
  # Plot site scores
  #geom_point(data = site_scores, aes(x = CAP1, y = CAP2, color = meta_df$Bottom_type), size = 3) +
  geom_jitter(data = site_scores, aes(x = CAP1, y = CAP2, color = meta_df$Bottom_type), 
            width = 0.07, height = 0.07, size = 3) +
  
  # Plot ellipses
  stat_ellipse(data = site_scores, aes(x = CAP1, y = CAP2, group = meta_df$Bottom_type, color = meta_df$Bottom_type), 
               level = 0.75, type = "norm", linetype = "solid", size = 1) + 
  
  # Plot arrows
  geom_segment(data = biplot_arrows, aes(x = 0, y = 0, xend = CAP1, yend = CAP2), 
               arrow = arrow(length = unit(0.3, "cm")), color = "black") + # Plot biplot arrows
  geom_text(data = biplot_arrows, aes(x = CAP1 * 1.1, y = CAP2 * 1.1, label = variable), color = "black") + # Label arrows
  
  # Plot species scores
  geom_point(data = species_scores_sig, aes(x = CAP1, y = CAP2), color = "black", size = 2) +
  geom_text_repel(data = species_scores_sig, aes(x = CAP1, y = CAP2, label = Species), color = "black", size = 3) +
  
  # Add labels and theme
  labs(x = paste("CAP1 (", round(proportion_variance[1], 2), "% variance explained)", sep = ""), 
       y = paste("CAP2 (", round(proportion_variance[2], 2), "% variance explained)", sep = ""),
       color = "bottom type") +
  theme_minimal() #+
  #theme(legend.position = "none")
plot_bottom 
```


```{r}
ggsave("/home/kimberly.ledger/goa-bt23-mb/figures/rkfish_rda.png", plot = plot_bottom, width = 8, height = 6, dpi = 300)
```


## what species are associated with longitude? 



